sycl/source/detail/graph_impl.cpp

```diff
@@ -0,0 +1,417 @@
+                                          // list
+    }
+  } else {
+    this->add_root(NodeImpl);
+  }
+
+  return NodeImpl;
+}
+
+std::shared_ptr<node_impl>
+graph_impl::add(const std::shared_ptr<graph_impl> &Impl,
+                std::function<void(handler &)> CGF,
+                const std::vector<sycl::detail::ArgDesc> &Args,
+                const std::vector<std::shared_ptr<node_impl>> &Dep) {
+  sycl::handler Handler{Impl};
+  CGF(Handler);
+  Handler.finalize();
+
+  // If the handler recorded a subgraph return that here as the relevant nodes
+  // have already been added. The node returned here is an empty node with
+  // dependencies on all the exit nodes of the subgraph.
+  if (Handler.MSubgraphNode) {
+    return Handler.MSubgraphNode;
+  }
+  if (Handler.MCGType == sycl::detail::CG::None) {
+    return this->add(Dep);
+  }
+  return this->add(Handler.MCGType, std::move(Handler.MGraphNodeCG), Dep);
+}
+
+std::shared_ptr<node_impl>
+graph_impl::add(sycl::detail::CG::CGTYPE CGType,
+                std::unique_ptr<sycl::detail::CG> CommandGroup,
+                const std::vector<std::shared_ptr<node_impl>> &Dep) {
+  // Copy deps so we can modify them
+  auto Deps = Dep;
+  if (CGType == sycl::detail::CG::Kernel) {
+    // A unique set of dependencies obtained by checking kernel arguments
+    // for accessors
+    std::set<std::shared_ptr<node_impl>> UniqueDeps;
+    const auto &Args =
+        static_cast<sycl::detail::CGExecKernel *>(CommandGroup.get())->MArgs;
+    for (auto &Arg : Args) {
+      if (Arg.MType != sycl::detail::kernel_param_kind_t::kind_accessor) {
+        continue;
+      }
+      // Look through the graph for nodes which share this argument
+      for (auto NodePtr : MRoots) {
+        check_for_arg(Arg, NodePtr, UniqueDeps);
+      }
+    }
+
+    // Add any deps determined from accessor arguments into the dependency list
+    Deps.insert(Deps.end(), UniqueDeps.begin(), UniqueDeps.end());
+  }
+
+  // Add any nodes specified by event dependencies into the dependency list
+  for (auto Dep : CommandGroup->getEvents()) {
+    if (auto NodeImpl = MEventsMap.find(Dep); NodeImpl != MEventsMap.end()) {
+      Deps.push_back(NodeImpl->second);
+    } else {
+      throw sycl::exception(errc::invalid,
+                            "Event dependency from handler::depends_on does "
+                            "not correspond to a node within the graph");
+    }
+  }
+
+  const std::shared_ptr<node_impl> &NodeImpl =
+      std::make_shared<node_impl>(CGType, std::move(CommandGroup));
+  if (!Deps.empty()) {
+    for (auto N : Deps) {
+      N->register_successor(NodeImpl, N); // register successor
+      this->remove_root(NodeImpl);        // remove receiver from root node
+                                          // list
+    }
+  } else {
+    this->add_root(NodeImpl);
+  }
+  return NodeImpl;
+}
+
+bool graph_impl::clear_queues() {
+  bool AnyQueuesCleared = false;
```

Comment:
It's [not something we've implemented now but are aware of as a gap](https://github.com/reble/llvm#implementation-status), so isn't pending to be added in another spit PR. If you have any advice on implementing this though it would be appreciated, and we could prioritize this task for follow-up work.